clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -494,6 +529,13 @@ public boolean hasUndrained() {
         if (nodes.isEmpty())
             return Collections.emptyMap();
 
+        // If we are past the current earliest delivery timeout time, just reset it.
+        // By the end of this function, we'll bring it down as we look at the batches.
+        // This ensures that earliest delivery timeout keeps moving ahead.
+        if (now >= earliestDeliveryTimeoutMs) {
+            earliestDeliveryTimeoutMs = Long.MAX_VALUE;
```

Comment:
After we reset `earliestDeliveryTimeoutMs`, it seems that we do not take into account the expiration times of in-flight batches.

Reply:
fixed. take a look.

Reply:
`RecordAccumulator.maybeUpdateEarliestDeliveryTimeout`